01.fit_a_line/train.py

```diff
@@ -13,122 +13,135 @@
-        if event.epoch % 10 == 0:
-            # We can save the trained parameters for the inferences later
+    main_program = fluid.default_main_program()
+    star_program = fluid.default_startup_program()
+
+    cost = fluid.layers.square_error_cost(input=y_predict, label=y)
+    avg_loss = fluid.layers.mean(cost)
+
+    sgd_optimizer = fluid.optimizer.SGD(learning_rate=0.001)
+    sgd_optimizer.minimize(avg_loss)
+
+    test_program = main_program.clone(for_test=True)
+
+    # can use CPU or GPU
+    use_cuda = False
+    place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
+    exe = fluid.Executor(place)
+
+    # Specify the directory to save the parameters
+    params_dirname = "fit_a_line.inference.model"
+    num_epochs = 100
+
+    # For training test cost
+    def train_test(program, feeder):
+        exe_test = fluid.Executor(place)
+        accumulated = 1 * [0]
+        count = 0
+        for data_test in test_reader():
+            outs = exe_test.run(
+                program=program,
+                feed=feeder.feed(data_test),
+                fetch_list=[avg_loss])
+            accumulated = [x_c[0] + x_c[1][0] for x_c in zip(accumulated, outs)]
+            count += 1
+        return [x_d / count for x_d in accumulated]
+
+    # main train loop.
+    def train_loop():
+
+        feeder = fluid.DataFeeder(place=place, feed_list=[x, y])
+        feeder_test = fluid.DataFeeder(place=place, feed_list=[x, y])
+        exe.run(star_program)
+
+        train_title = "Train cost"
+        test_title = "Test cost"
+        step = 0
+
+        for pass_id in range(num_epochs):
+            for data_train in train_reader():
+                avg_loss_value, = exe.run(
+                    main_program,
+                    feed=feeder.feed(data_train),
+                    fetch_list=[avg_loss])
+                if step % 10 == 0:  # record a train cost every 10 batches
+                    print("%s, Step %d, Cost %f" %
+                          (train_title, step, avg_loss_value[0]))
+                if step % 100 == 0:  # record a test cost every 100 batches
+                    test_metics = train_test(
+                        program=test_program, feeder=feeder_test)
+                    print("%s, Step %d, Cost %f" %
+                          (test_title, step, test_metics[0]))
+                    # If the accuracy is good enough, we can stop the training.
+                    if test_metics[0] < 10.0:
+                        return
+
+                step += 1
+
+                if math.isnan(float(avg_loss_value)):
+                    sys.exit("got NaN loss, training failed.")
             if params_dirname is not None:
-                trainer.save_params(params_dirname)
+                # We can save the trained parameters for the inferences later
+                fluid.io.save_inference_model(params_dirname, ['x'],
+                                              [y_predict], exe)
 
+    train_loop()
```

Comment:
`train_loop` is not needed. We can in-place it right here.

Reply:
ok，done